File: src/main/java/com/example/misyemek/Service/SiparisService.java
package com.example.misyemek.Service;

import com.example.misyemek.entity.Firma;
import com.example.misyemek.entity.Siparis;
import com.example.misyemek.repository.FirmaRepository;
import com.example.misyemek.repository.SiparisRepository;
import org.springframework.stereotype.Service;
import com.example.misyemek.Enum.SiparisDurumu;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;

import static com.example.misyemek.Enum.SiparisDurumu.*;

@Service
public class SiparisService {
    private final SiparisRepository repository;
    private final FirmaRepository firmaRepository;

    public SiparisService(SiparisRepository repository, FirmaRepository firmaRepository) {
        this.repository = repository;
        this.firmaRepository = firmaRepository;
    }

    public Siparis siparisEkle(Siparis yeniSiparis) {
        return repository.save(yeniSiparis);
    }

    public List<Object[]> kullaniciIdSiparisGetir(Long kullaniciId) {
        return repository.kullaniciIdSiparisGetir(kullaniciId);

    }
    public List<Object[]> firmaIdSiparisGetir(Long firmaId){
        return  repository.firmaIdSiparisGetir(firmaId);
    }

    public List<Siparis> siparisGetir(Long kullaniciId) {
        return repository.findByKullaniciId(kullaniciId);

    }

    public void siparisSil(Long siparisId) {
        if (!repository.existsById(siparisId)) {
            throw new RuntimeException("sipariş bulunamadı" + siparisId);
        }
        repository.deleteById(siparisId);
    }

    public BigDecimal toplamHarcama(Long kullaniciId, Long firmaId){
        return repository.toplamHarcama(kullaniciId, firmaId);
    }

    public BigDecimal toplamCiro(Long firmaId) {
        return repository.toplamCiro(firmaId);
    }

    public List<Siparis> sepetGetir(Long kullaniciId){
        return repository.findByKullaniciIdAndSiparisDurumu(kullaniciId, SEPETTE);}

    @Transactional
    public List<Siparis> sepetGuncelle(Long kullaniciId){
        List<Siparis> sepet = repository.findByKullaniciIdAndSiparisDurumu(kullaniciId, SEPETTE);
        int grupNo = repository.maxGrupNo() + 1;
        for (Siparis s : sepet) {
            s.setSiparisDurumu(SiparisDurumu.HAZIRLANIYOR);
            s.setGrupNo(grupNo);
        }

        return repository.saveAll(sepet);
    }

    public Siparis sepetOnayla(Long siparisId){
        Siparis siparisler = repository.findById(siparisId)
                .orElseThrow(() -> new RuntimeException("kullanıcı ıd bulunamadı") );
            siparisler.setSiparisDurumu(SiparisDurumu.YOLDA);
        return repository.save(siparisler);
    }

    public Siparis sepetReddet(Long kullaniciId){
    Siparis siparisler =repository.findById(kullaniciId)
                    .orElseThrow(()-> new RuntimeException("kullanici bulunamadı"));
    siparisler.setSiparisDurumu(SiparisDurumu.IPTAL);
    return repository.save(siparisler);
    }

    public List<Object[]> kuryeSiparis(){
        return repository.kuryeSiparis();
    }

    @Transactional
    public List<Siparis> kuryeSiparisTeslimEdildi(int grupNo , Long firmaId){
        List<Siparis> siparisler = repository.findByGrupNoAndFirmaUrun_FirmaId(grupNo ,firmaId);
    for (Siparis s : siparisler) {
        s.setSiparisDurumu(SiparisDurumu.TESLIM_EDILDI);
    }
        return repository.saveAll(siparisler);
    }

    public List<Siparis> kuryeSiparisReddet(int grupNo , Long firmaId){
        List<Siparis> siparis = repository.findByGrupNoAndFirmaUrun_FirmaId(grupNo ,firmaId);
        for (Siparis s : siparis) {
            s.setSiparisDurumu(SiparisDurumu.IPTAL);
        }
        return repository.saveAll(siparis);
    }

    public List<Object[]> firmaSiparisTakip (Long firmaId){
        return repository.firmaSiparisTakip(firmaId);
    }

    public Long firmaTeslimSayac(Long firmaId ,int yil , int ay ){
        return repository.firmaTeslimSayac(firmaId,yil,ay);
    }

    public Long firmaIptalSayac(Long firmaId  ,int yil , int ay ){
        return repository.firmaIptalSayac(firmaId, yil , ay);
    }

    public Double yildizSay(Long firmaId){
        return repository.yildizSay(firmaId);
    }

    @Transactional
    public Siparis yildizEkle(Long firmaId, int grupNo, int yildizSayisi){
        List<Siparis> siparisler = repository.findByGrupNoAndFirmaUrun_FirmaId(grupNo, firmaId);

        if (siparisler.isEmpty()) {
            throw new RuntimeException("sipariş bulunamadı");
        }
        Siparis ilk = siparisler.get(0);
        ilk.setYildizSayisi(yildizSayisi);
        repository.save(ilk);
        Double ortalama =repository.yildizSay(firmaId);
        Firma firma = firmaRepository.findById(firmaId)
                .orElseThrow(()-> new RuntimeException("firma bulunamadı"));
        firma.setPuanOrtalama(ortalama);
        return ilk;
    }
    }

File: frontend/src/app/Service/siparis.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { BaseService } from '../core/services/base.service';

    @Injectable({providedIn: 'root'})
  export class SiparisService extends BaseService {
  protected baseUrl = '/Siparisler';

  sepetGetir(): Observable<any[]> {
  return this.get<any[]>(`Sepet/${this.kullaniciId}`);
  }

  kullaniciIdSiparisGetir(): Observable<any[]> {
  return this.get<any[]>(`Getir/${this.kullaniciId}`);
  }
  
  sepetEkle(siparis: any): Observable<any>{
    return this.post<any>('Kaydet', siparis)
  }

  sepetSil(siparisId: number): Observable<any[]>{
    return this.delete<any[]>(`Sil/${siparisId}`)
  }

  sepetOde(): Observable<any[]>{
    return this.put<any[]>(`Sepet/Ode/${this.kullaniciId}`, {})
  }

  firmaIdSiparisGetir(firmaId: number | string): Observable<any[]>{
    return this.get<any[]>(`FirmaGetir/${firmaId}`)
  }

  sepetOnayla(siparisId: number): Observable<any>{
    return this.put<any>(`Sepet/Onayla/${siparisId}`,{})
  }

  sepetReddet(siparisId: number): Observable<any>{
    return this.put<any>(`Sepet/Reddet/${siparisId}`, {})
  }

  kuryeSiparis(): Observable<any[]>{
    return this.get<any[]>(`Kurye/Siparis`)
  }

  kuryeSiparisOnayla(firmaId: number , grupNo: number ): Observable<any>{
    return this.put<any>(`Kurye/TeslimEdildi/${firmaId}/${grupNo}` , {})
  }

  kuryeSiparisReddet(firmaId: number , grupNo: number): Observable<any>{
    return this.put<any>(`Kurye/Reddet/${firmaId}/${grupNo}` , {})
  } 

  sepetTakip(): Observable<any[]>{
    return this.get<any[]>(`${this.kullaniciId}`)
  }

  sepetTakipFirma(firmaId: number): Observable<any[]>{
    return this.get<any[]>(`Firma/SiparisTakip/${firmaId}`)
  }

  firmaTeslimSayac(firmaId: string , yil: number , ay: number): Observable<any>{
    return this.get<any>(`TeslimEdildiSayac/${firmaId}/${yil}/${ay}`)
  }

  firmaIptalSayac(firmaId: string , yil: number , ay: number): Observable<any>{
    return this.get<any>(`IptalSayac/${firmaId}/${yil}/${ay}`)
  }

  yildizSay(firmaId: string): Observable <any>{
    return this.get<any>(`Firma/YildizSayisi/${firmaId}`)
  }

  yildizEkle(firmaId: number , grupNo: number , yildizSayisi: number): Observable<any>{
    return this.put<any>(`Firma/YildizEkle/${firmaId}/${grupNo}/${yildizSayisi}`,{})
  }

}